clang/runtime/dpct-rt/include/sparse_utils.hpp.inc

```diff
@@ -78,6 +79,241 @@ private:
   oneapi::mkl::index_base _index_base = oneapi::mkl::index_base::zero;
 };
 // DPCT_LABEL_END
+
+// DPCT_LABEL_BEGIN|csrmv|dpct::sparse
+// DPCT_DEPENDENCY_BEGIN
+// SparseUtils|matrix_info
+// Util|DataType
+// LibCommonUtils|get_value
+// LibCommonUtils|get_memory
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+/// Computes a CSR format sparse matrix-dense vector product.
+/// y = alpha * op(A) * x + beta * y
+/// \param [in] queue The queue where the routine should be executed.
+/// \param [in] trans The operation applied to the matrix A.
+/// \param [in] num_rows Number of rows of the matrix A.
+/// \param [in] num_cols Number of columns of the matrix A.
+/// \param [in] alpha Scaling factor for the matrix A.
+/// \param [in] info Matrix info of the matrix A.
+/// \param [in] val An array containing the non-zero elements of the matrix A.
+/// \param [in] row_ptr An array of length \p num_rows + 1.
+/// \param [in] col_ind An array containing the column indices in index-based
+/// numbering.
+/// \param [in] x Data of the vector x.
+/// \param [in] beta Scaling factor for the vector x.
+/// \param [in, out] y Data of the vector y.
+template <typename T>
+void csrmv(sycl::queue &queue, oneapi::mkl::transpose trans, int num_rows,
+           int num_cols, const T *alpha,
+           const std::shared_ptr<matrix_info> info, const T *val,
+           const int *row_ptr, const int *col_ind, const T *x, const T *beta,
+           T *y) {
+  using Ty = typename dpct::DataType<T>::T2;
+  auto alpha_value =
+      detail::get_value(reinterpret_cast<const Ty *>(alpha), queue);
+  auto beta_value =
+      detail::get_value(reinterpret_cast<const Ty *>(beta), queue);
+
+  oneapi::mkl::sparse::matrix_handle_t *sparse_matrix_handle =
+      new oneapi::mkl::sparse::matrix_handle_t;
```

Comment:
```suggestion
  oneapi::mkl::sparse::matrix_handle_t sparse_matrix_handle = nullptr;
```

Reply:
I would recommend to use `matrix_handle_t sparse_matrix_handle` instead of `matrix_handle_t *p_sparse_matrix_handle`.  All the below suggestions will align with this usage recommendation.  In particular it removes the need for the last host task to call delete.

Reply:
The `release_matrix_handle` is async and it accpets the address of `sparse_matrix_handle`.
If we use `sparse_matrix_handle` directly, will the address of `sparse_matrix_handle` be invalid after this function returned?

Reply:
sparse_matrix_handle is a pointer to object.  It is a pointer to no object (nullptr) when first declared, then taht object is created on init_matrix_handle and the address the pointer is pointing to is updated to point to the newly created object -- think of init_matrix_handle as "new".  Likewise that object is destroyed in release_matrix_handle (think of it as "delete") and the pointer is updated to point to nullptr again.  The construction (init_matrix_handle) is sychronous but the destruction (release_matrix_handle) is asynchronous.

Reply:
E.g.,
```
  1 void foo() {
  2   oneapi::mkl::sparse::matrix_handle_t handle;
  3   oneapi::mkl::sparse::init_matrix_handle(&handle);
  4   ...
  5   oneapi::mkl::sparse::release_matrix_handle(queue, &handle, ...);
  6 }
  7 
  8 int main() {
  9   ...
 10   foo();
 11   printf("123\n");
 12   ...
 13 }
```
When the program goes to Line11, at this time, the variable `handle` has been de-constructed, so the address of that variable `&handle` , which passed to `release_matrix_handle`, is invalid.
The issue here is that the function `release_matrix_handle` accepts `matrix_handle_t *` but not `matrix_handle_t`, so I need to make sure the address of the handle is valid until `release_matrix_handle` is done.

Reply:
Maybe in your implementation of the `release_matrix_handle`, you dereference the input `&handle` and save it somewhere before the function return. Then your suggestion can work well. But from the interface, since it accepts a pointer, I think using new/delete is safer.